### PR TITLE
sub AckPolicy and description

### PR DESCRIPTION
**Pull request by roeschter**
Improved description. Added support for ackPolicy

### CODE REVIEW

File: cli/sub_command.go [hunk 1]
```diff
@@ -73,6 +76,7 @@ func configureSubCommand(app commandHost) {
 	act.Flag("raw", "Show the raw data received").Short('r').UnNegatableBoolVar(&c.raw)
 	act.Flag("translate", "Translate the message data by running it through the given command before output").StringVar(&c.translate)
 	act.Flag("ack", "Acknowledge JetStream message that have the correct metadata").BoolVar(&c.jsAck)
+	act.Flag("ackPolicy", "Acknowledgment policy (none, all, explicit) (requires JetStream)").Default("none").StringVar(&c.ackPolicy)
```
[the remark this answers is not in the document]
Comment (ripienaar) [reply cli/sub_command.go:79]: I wonder though whats the utility of supporting like AckAll? Or Explicit? For the CLI `sub` command what would be the different we ack every message or no messages, no need to differentiate between all and explicit I think?

File: cli/sub_command.go [hunk 2]
```diff
@@ -451,7 +471,7 @@ func (c *subCmd) subscribe(p *fisk.ParseContext) error {
 			}
 			subs = append(subs, sub)
 		} else {
-			c.jsAck = false
+
```
Comment (ripienaar) [inline cli/sub_command.go:474]: If we arent binding to a specific durable, we make an ephemeral consumer and there's no need for acks. Why ack it? We just show it on the stream once and thats it, hence for the `sub` behaviour being just "tail a stream" no acks are completely fine.

We only need to ack things when we bind to a consumer that has ack set, this is why its always forced off here

I guess I am just wondering why we need to support all the ack behaviours at all here

**Review by ripienaar — CHANGES_REQUESTED** [review]

File: cli/sub_command.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (roeschter) [reply cli/sub_command.go:79]: Late reply, was very busy this week. 
I was experimenting and performance testing the behavior of QueuePolicy streams. While it doesn't matter what happens to the message in all other cases, for QueuePolicy  only an ack will remove the message. Explicit ack is required in this case. AckAll was added just for completeness sake.

File: cli/sub_command.go (hunk 2 above)
Comment (roeschter) [reply cli/sub_command.go:474]: Special case for QueuePolicy (see previous comment). For all other retention policies it does not matter indeed.

File: cli/sub_command.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ripienaar) [reply cli/sub_command.go:79]: In that case we do not need to set the particular policy.  If --ack is given then do explicit ack when creating the consumer thats it, works in all cases even WQ.

[the remark this answers is not in the document]
Comment (ripienaar) [reply cli/sub_command.go:79]: I am just mindful of option bloat, cli - and especially this command - has a ton of options, so I try to think before adding any.  There's no use in adding none,all,explicit when none is the default and all has no use in this case.  That leaves explicit which `--ack` implies.

**Comment by ripienaar** [comment]
```
diff --git a/cli/sub_command.go b/cli/sub_command.go
index 98c59d6..0f99f93 100644
--- a/cli/sub_command.go
+++ b/cli/sub_command.go
@@ -451,7 +451,10 @@ func (c *subCmd) subscribe(p *fisk.ParseContext) error {
 			}
 			subs = append(subs, sub)
 		} else {
-			c.jsAck = false
+			if c.jsAck {
+				opts = append(opts, nats.AckExplicit())
+			}
+
 			sub, err := js.Subscribe(c.firstSubject(), handler, opts...)
 			if err != nil {
 				return err

```

is enough to make this work imo.  When accessing a WQ it will error by default

```
$ nats sub --stream WQ
nats: error: nats: workqueue stream requires explicit ack
```

and then you pass `--ack` to make it work, I like the error since ack on a WQ is destructive and so I prefer people opt into that behaviour explicitely.

**Comment by ripienaar** [comment]
Also worth noting it also doesnt work if there's a consumer on the WQ since there would be overlap.  Personally I think not working against WQ is a feature - too easy to destroy all messages in the WQ with a simple fat finger.

**Comment by roeschter** [comment]
Will not implement those changes, could accidentally delete QueuePolicy content. Adding to help for the time being

**Pull request closed without merging**